src/conn/conn_compact.c

```diff
@@ -0,0 +1,270 @@
+        /*
+         * Compact may return:
+         * - EBUSY for various reasons.
+         * - ETIMEDOUT if the timer has been configured and compaction took too long.
+         * - WT_NOTFOUND if the underlying file has been deleted.
+         */
+        if (ret == EBUSY || ret == WT_NOTFOUND || ret == ETIMEDOUT)
+            ret = 0;
+        WT_ERR(ret);
+    }
+
+    WT_STAT_CONN_SET(session, session_background_compact_running, 0);
+
+    WT_ERR(__wt_metadata_cursor_close(session));
+    __wt_free(session, config);
+    __wt_free(session, conn->background_compact.config);
+
+    if (0) {
+err:
+        WT_IGNORE_RET(__wt_panic(session, ret, "compact server error"));
+    }
+    return (WT_THREAD_RET_VALUE);
+}
+
+/*
+ * __wt_compact_server_create --
+ *     Start the compact thread.
+ */
+int
+__wt_compact_server_create(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    uint32_t session_flags;
+
+    conn = S2C(session);
+
+    /* Set first, the thread might run before we finish up. */
+    FLD_SET(conn->server_flags, WT_CONN_SERVER_COMPACT);
+
+    /*
+     * Compaction does enough I/O it may be called upon to perform slow operations for the block
+     * manager.
+     */
+    session_flags = WT_SESSION_CAN_WAIT;
+    WT_RET(__wt_open_internal_session(
+      conn, "compact-server", true, session_flags, 0, &conn->background_compact.session));
+    session = conn->background_compact.session;
+
+    WT_RET(__wt_cond_alloc(session, "compact server", &conn->background_compact.cond));
+
+    WT_RET(__wt_thread_create(session, &conn->background_compact.tid, __compact_server, session));
+    conn->background_compact.tid_set = true;
+
+    return (0);
+}
+
+/*
+ * __wt_compact_server_destroy --
+ *     Destroy the background compaction server thread.
+ */
+int
+__wt_compact_server_destroy(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+
+    conn = S2C(session);
+
+    FLD_CLR(conn->server_flags, WT_CONN_SERVER_COMPACT);
+    if (conn->background_compact.tid_set) {
+        conn->background_compact.running = false;
+        __wt_cond_signal(session, conn->background_compact.cond);
+        WT_TRET(__wt_thread_join(session, &conn->background_compact.tid));
+        conn->background_compact.tid_set = false;
+    }
+    __wt_cond_destroy(session, &conn->background_compact.cond);
+
+    /* Close the server thread's session. */
+    if (conn->background_compact.session != NULL) {
+        WT_TRET(__wt_session_close_internal(conn->background_compact.session));
+        conn->background_compact.session = NULL;
+    }
+
+    return (ret);
+}
+
+/*
+ * __wt_compact_signal --
+ *     Signal the compact thread. Return an error if the background compaction server has not
+ * processed a previous signal yet or because of an invalid configuration.
+ */
+int
+__wt_compact_signal(WT_SESSION_IMPL *session, const char *config)
```

Comment:
What if an application calls this function while WiredTiger is opening/closing? Is that undefined behaviour? It seems that a race is possible.

Reply:
@keitharnoldsmith, @kommiharibabu is it assumed that an application should not do anything during reconfiguration/open/close? Or should I try to make sure we are robust against this kind of scenario?

Reply:
WT doesn't allow any application sessions to be opened during open/close. The scenario you are mentioning shouldn't occur. Connection reconfiguration is where we didn't close the sessions, do we need to something for background compact in `__wt_conn_reconfig` function?

Reply:
After double checking the code, yes it should be safe during open/close.
Regarding reconfiguration, I don't think there is any harm to enable/disable the background compaction server at the same time.

Reply:
In general, we should assume the application will only do legal things. So we should be robust against this kind of thing. Mostly we should make sure that it can't cause a crash or other bad behavior. 

I'm not sure how this function is going to be connected to the API, but if it is something that it triggered via config options on the `WT_SESSION.compact` call, I suspect this will "just work" because we must already handle races with calling compact. But it would be worth looking at the code a bit to confirm.